tsx file compile fails on optional properties

Once reported here, https://github.com/TypeStrong/ntypescript/issues/17 but it seems that this problem happens on the current typescript trunk version.

`tsconfig.json`:
````
{
  "compilerOptions": {
    "target": "es5",
    "module": "amd",
    "jsx": "react"
  },
  "compileOnSave": false,
  "files": [
    "./main.tsx",
    "./typings/tsd.d.ts"
  ]
}
````

`tsd`:
````
tsd install react-global react-bootstrap --save 
````

`main.tsx`:
````
import * as React from 'react';
import { render } from 'react-dom';
import { Button } from 'react-bootstrap';

class App extends React.Component<{}, {}> {
  render() {
    return <Button bsSize="large" bsStyle="danger" > {"Button Text"} </Button>;
  }
}

render(<App />, document.getElementsByTagName("entrypoint")[0] );
````

`typescript 1.8.0-dev.20151210` tsc fails with these error messages
````
main.tsx(8,12): error TS2324: Property 'key' is missing in type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<ButtonProps, any>> & ButtonProps'.
main.tsx(8,12): error TS2324: Property 'ref' is missing in type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<ButtonProps, any>> & ButtonProps'.
````
The error message says that 'key' and 'props' property must be given, but they are defined as optional on IntrinsicAttributes and IntrinsicClassicAttributes like this. 

in `react.d.ts`, line 2119~2125
````
    interface IntrinsicAttributes {
        key?: string | number;
    }

    interface IntrinsicClassAttributes<T> {
        ref?: string | ((classInstance: T) => void);
    }
````
`typescript 1.7.3` compile succeeds without problem.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -3766,13 +3766,14 @@ namespace ts {
         function createUnionOrIntersectionProperty(containingType: UnionOrIntersectionType, name: string): Symbol {
             const types = containingType.types;
             let props: Symbol[];
-            let isOptional = !!(containingType.flags & TypeFlags.Intersection);
+            // Flags we want to propagate to the result if they exist in all source symbols
+            let commonFlags = (containingType.flags & TypeFlags.Intersection) ? SymbolFlags.Optional : SymbolFlags.None;
             for (const current of types) {
                 const type = getApparentType(current);
                 if (type !== unknownType) {
                     const prop = getPropertyOfType(type, name);
                     if (prop && !(getDeclarationFlagsFromSymbol(prop) & (NodeFlags.Private | NodeFlags.Protected))) {
-                        isOptional = isOptional && !!(prop.flags & SymbolFlags.Optional);
+                        commonFlags &= prop.flags;
                         if (!props) {
                             props = [prop];
                         }
@@ -3804,7 +3805,7 @@ namespace ts {
                 SymbolFlags.Property |
                 SymbolFlags.Transient |
                 SymbolFlags.SyntheticProperty |
-                (isOptional ? SymbolFlags.Optional : SymbolFlags.None),
+                commonFlags,
                 name);
             result.containingType = containingType;
             result.declarations = declarations;

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -3766,11 +3766,13 @@ namespace ts {
         function createUnionOrIntersectionProperty(containingType: UnionOrIntersectionType, name: string): Symbol {
             const types = containingType.types;
             let props: Symbol[];
+            let isOptional = !!(containingType.flags & TypeFlags.Intersection);
             for (const current of types) {
                 const type = getApparentType(current);
                 if (type !== unknownType) {
                     const prop = getPropertyOfType(type, name);
                     if (prop && !(getDeclarationFlagsFromSymbol(prop) & (NodeFlags.Private | NodeFlags.Protected))) {
+                        isOptional = isOptional && !!(prop.flags & SymbolFlags.Optional);
                         if (!props) {
                             props = [prop];
                         }
@@ -3798,7 +3800,12 @@ namespace ts {
                 }
                 propTypes.push(getTypeOfSymbol(prop));
             }
-            const result = <TransientSymbol>createSymbol(SymbolFlags.Property | SymbolFlags.Transient | SymbolFlags.SyntheticProperty, name);
+            const result = <TransientSymbol>createSymbol(
+                SymbolFlags.Property |
+                SymbolFlags.Transient |
+                SymbolFlags.SyntheticProperty |
+                (isOptional ? SymbolFlags.Optional : SymbolFlags.None),
+                name);
             result.containingType = containingType;
             result.declarations = declarations;
             result.type = containingType.flags & TypeFlags.Union ? getUnionType(propTypes) : getIntersectionType(propTypes);
@@ -8232,9 +8239,9 @@ namespace ts {
                             // Props is of type 'any' or unknown
                             return links.resolvedJsxType = attributesType;
                         }
-                        else if (!(attributesType.flags & TypeFlags.ObjectType)) {
-                            // Props is not an object type
-                            error(node.tagName, Diagnostics.JSX_element_attributes_type_0_must_be_an_object_type, typeToString(attributesType));
+                        else if (attributesType.flags & TypeFlags.Union) {
+                            // Props cannot be a union type
+                            error(node.tagName, Diagnostics.JSX_element_attributes_type_0_may_not_be_a_union_type, typeToString(attributesType));
                             return links.resolvedJsxType = anyType;
                         }
                         else {

--- a/src/compiler/diagnosticMessages.json
+++ b/src/compiler/diagnosticMessages.json
@@ -1691,7 +1691,7 @@
         "category": "Error",
         "code": 2528
     },
-    "JSX element attributes type '{0}' must be an object type.": {
+    "JSX element attributes type '{0}' may not be a union type.": {
         "category": "Error",
         "code": 2600
     },

--- a/tests/baselines/reference/tsxAttributeResolution11.errors.txt
+++ b/tests/baselines/reference/tsxAttributeResolution11.errors.txt
@@ -1,0 +1,33 @@
+tests/cases/conformance/jsx/file.tsx(11,22): error TS2339: Property 'bar' does not exist on type 'IntrinsicAttributes & { ref?: string; }'.
+
+
+==== tests/cases/conformance/jsx/react.d.ts (0 errors) ====
+    
+    declare module JSX {
+    	interface Element { }
+    	interface IntrinsicElements {
+    	}
+    	interface ElementAttributesProperty {
+    		props;
+    	}
+    	interface IntrinsicAttributes {
+    		ref?: string;
+    	}
+    }
+    
+==== tests/cases/conformance/jsx/file.tsx (1 errors) ====
+    class MyComponent {  
+      render() {
+      }
+    
+      props: {
+    	  ref?: string;
+      }
+    }
+    
+    // Should be an OK
+    var x = <MyComponent bar='world' />;
+                         ~~~
+!!! error TS2339: Property 'bar' does not exist on type 'IntrinsicAttributes & { ref?: string; }'.
+    
+    

--- a/tests/baselines/reference/tsxAttributeResolution11.js
+++ b/tests/baselines/reference/tsxAttributeResolution11.js
@@ -1,0 +1,41 @@
+//// [tests/cases/conformance/jsx/tsxAttributeResolution11.tsx] ////
+
+//// [react.d.ts]
+
+declare module JSX {
+	interface Element { }
+	interface IntrinsicElements {
+	}
+	interface ElementAttributesProperty {
+		props;
+	}
+	interface IntrinsicAttributes {
+		ref?: string;
+	}
+}
+
+//// [file.tsx]
+class MyComponent {  
+  render() {
+  }
+
+  props: {
+	  ref?: string;
+  }
+}
+
+// Should be an OK
+var x = <MyComponent bar='world' />;
+
+
+
+//// [file.jsx]
+var MyComponent = (function () {
+    function MyComponent() {
+    }
+    MyComponent.prototype.render = function () {
+    };
+    return MyComponent;
+}());
+// Should be an OK
+var x = <MyComponent bar='world'/>;

--- a/tests/cases/conformance/jsx/tsxAttributeResolution11.tsx
+++ b/tests/cases/conformance/jsx/tsxAttributeResolution11.tsx
@@ -1,0 +1,29 @@
+//@jsx: preserve
+//@module: amd
+
+//@filename: react.d.ts
+declare module JSX {
+	interface Element { }
+	interface IntrinsicElements {
+	}
+	interface ElementAttributesProperty {
+		props;
+	}
+	interface IntrinsicAttributes {
+		ref?: string;
+	}
+}
+
+//@filename: file.tsx
+class MyComponent {  
+  render() {
+  }
+
+  props: {
+	  ref?: string;
+  }
+}
+
+// Should be an OK
+var x = <MyComponent bar='world' />;
+

